### ISSUE TITLE
Segmentation fault on Mac OS X Lion involving shared variable

### ISSUE THREAD

**Issue by bayerj** [Theano/Theano#24]
This code gives a segmentation fault during compilation on the following setup:
- Recent version of scipy superpack
- current HEAD (15e5ae65b14a7a442174) of theano as well as tag rel-0.4.1
- all other libraries are the preinstalled ones from Apple

This fault does not happen on the same setup with Max OS 10.6.

``` python
import scipy
import theano
import theano.tensor as T
from zeitgeist.util import ParameterSet

w = theano.shared(scipy.eye(2))
inpt = T.matrix()
transformed = T.dot(inpt, w)
energy = transformed.sum()

energyprime = T.grad(energy, inpt)
loss = (0.5 * energyprime**2).sum()

f = theano.function([inpt], loss)
i = ((1, 2), (3, 4))
print f(i)
```

There are two workarounds for this:
- not using a shared variable for w,
- pulling the 0.5 factor in the loss expression out of the sum.

The same happens when using the gradient of a related model:

``` python

import scipy
import theano
import theano.tensor as T
from zeitgeist.util import ParameterSet

w = T.matrix('weights')
w = theano.shared(scipy.eye(2))
inpt = T.matrix()
transformed = T.dot(inpt, w)
transformed = T.log(1 + 0.5 * transformed**2)
energy = transformed.sum()

energyprime = T.grad(energy, inpt)
loss = 0.5 * (energyprime**2).sum()

lossgrad = T.grad(loss, w)

i = ((1, 2), (3, 4))
f = theano.function([inpt], lossgrad)
print f(i)
```

The major difference is in the additional parts around the `transformed` variable. I have not found a workaround for this.
